nselib/idna.lua

```diff
@@ -0,0 +1,614 @@
+-- @param checkHyphens Boolean flag checks for 0x002D in unusual places.
+function validate(tableOfTables, checkHyphens)
+
+  if checkHyphens == nil then
+    checkHyphens = true
+  end
+
+  -- Validates the list of input codepoints.
+  for _, tbl in ipairs(tableOfTables) do
+
+    if checkHyphens then
+
+      -- Checks the 3rd and 4th position of input.
+      if (tbl[3] and tbl[3] == 0x002D) or (tbl[4] and tbl[4] == 0x002D) then
+        return false
+      end
+
+      -- Checks for starting and ending of input.
+      if tbl[1] == 0x002D or tbl[#tbl] == 0x002D then
+        return false
+      end
+
+    end
+
+    for _, v in ipairs(tbl) do
+      if v == 0x002E then
+        return false
+      end
+    end
+
+    -- TODO:
+    -- 1. Add validation for checkBidi, checkJoiners (if required).
+    -- 2. The label must not begin with a combining mark, that is: General_Category=Mark.
+  end
+
+  return true
+
+end
+
+-- This function converts the input codepoints into ASCII text based on IDNA rules.
+--
+-- @param codepoints Table of codepoints of decoded input.
+-- @param tbl Table of optional params.
+-- @param checkHyphens Boolean flag for checking hyphens presence in input.
+-- @param checkBidi Boolean flag to represent if the input is of Bidi type.
+-- @param checkJoiners Boolean flag to check for ContextJ rules in input.
+-- @param useSTD3ASCIIRules Boolean value to represent ASCII rules.
+-- @param transitionalProcessing Boolean value.
+-- @param delimiter codepoint of the character to be used as delimiter.
+-- @param encoder Encoder function to convert a Unicode codepoint into a
+-- string of bytes.
+-- @param An decoder function to decode the input string
+-- into an array of code points.
+-- @return Returns the IDNA ASCII format of the input.
+-- @return Throws nil, if there is any error in conversion.
+function toASCII(codepoints, transitionalProcessing, checkHyphens, checkBidi, checkJoiners, useSTD3ASCIIRules, delimiter, encoder, decoder)
+
+  -- Assigns default values if not specified.
+  if transitionalProcessing == nil then
+    transitionalProcessing = true
+  end
+  if checkHyphens == nil then
+    checkHyphens = true
+  end
+
+  -- Bidi refers to right-to-left scripts.
+  -- Labels must satisfy all six of the numbered conditions in RFC 5893, Section 2.
+  -- to use checkBidi functionality.
+  if checkBidi == nil then
+    checkBidi = false
+  end
+
+  -- Labels must satisify the ContextJ rules to use checkJoiners functionality.
+  if checkJoiners == nil then
+    checkJoiners = false
+  end
+
+  if useSTD3ASCIIRules == nil then
+    useSTD3ASCIIRules = true
+  end
+
+  delimiter = delimiter or 0x002E
+  encoder = encoder or unicode.utf8_enc
+  decoder = decoder or unicode.utf8_dec
+
+  local inputString = unicode.encode(codepoints, encoder)
+
+  -- Checks for invalid domain codepoints and proceeds further.
+  if not match(inputString, "[^a-z0-9_@=+/%-%(%)%%]") then
```

Comment:
What's the reference for these valid code points/characters? I'm surprised to see `_@=+()%` in there.

Reply:
My bad, I deleted this piece of code.